ampd/src/handlers/starknet_verify_msg.rs

```diff
@@ -0,0 +1,655 @@
+    pub source_address: CheckedFelt,
+    pub payload_hash: Hash,
+}
+
+#[derive(Deserialize, Debug)]
+#[try_from("wasm-messages_poll_started")]
+struct PollStartedEvent {
+    #[serde(rename = "_contract_address")]
+    contract_address: TMAddress,
+    poll_id: PollId,
+    source_gateway_address: String,
+    expires_at: u64,
+    messages: Vec<Message>,
+    participants: Vec<TMAddress>,
+}
+
+pub struct Handler<C>
+where
+    C: StarknetClient,
+{
+    verifier: TMAddress,
+    voting_verifier: TMAddress,
+    rpc_client: C,
+    latest_block_height: Receiver<u64>,
+}
+
+impl<C> Handler<C>
+where
+    C: StarknetClient + Send + Sync,
+{
+    pub fn new(
+        verifier: TMAddress,
+        voting_verifier: TMAddress,
+        rpc_client: C,
+        latest_block_height: Receiver<u64>,
+    ) -> Self {
+        Self {
+            verifier,
+            voting_verifier,
+            rpc_client,
+            latest_block_height,
+        }
+    }
+
+    fn vote_msg(&self, poll_id: PollId, votes: Vec<Vote>) -> MsgExecuteContract {
+        MsgExecuteContract {
+            sender: self.verifier.as_ref().clone(),
+            contract: self.voting_verifier.as_ref().clone(),
+            msg: serde_json::to_vec(&ExecuteMsg::Vote { poll_id, votes })
+                .expect("vote msg should serialize"),
+            funds: vec![],
+        }
+    }
+}
+
+#[async_trait]
+impl<V> EventHandler for Handler<V>
+where
+    V: StarknetClient + Send + Sync,
+{
+    type Err = Error;
+
+    async fn handle(&self, event: &events::Event) -> Result<Vec<Any>> {
+        let PollStartedEvent {
+            poll_id,
+            source_gateway_address,
+            messages,
+            participants,
+            expires_at,
+            contract_address,
+            ..
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(vec![]);
+            }
+            event => event.change_context(DeserializeEvent)?,
+        };
+
+        if self.voting_verifier != contract_address {
+            return Ok(vec![]);
+        }
+
+        if !participants.contains(&self.verifier) {
+            return Ok(vec![]);
+        }
+
+        let latest_block_height = *self.latest_block_height.borrow();
+        if latest_block_height >= expires_at {
+            info!(poll_id = poll_id.to_string(), "skipping expired poll");
+            return Ok(vec![]);
+        }
+
+        let unique_msgs = messages
+            .iter()
+            .unique_by(|msg| msg.message_id.to_string())
+            .collect::<Vec<_>>();
+
+        // key is the message_id of the tx holding the event
+        let mut events: HashMap<FieldElementAndEventIndex, ContractCallEvent> =
+            try_join_all(unique_msgs.iter().map(|msg| {
+                self.rpc_client
+                    .get_events_by_hash_contract_call(msg.message_id.tx_hash.clone())
```

Comment:
why not pass in the event index as well, and just return the event if found? This `events` map is going to have a ton of events that the code doesn't care about, would probably be more efficient and clearer to not include those

Reply:
This `get_events_by_hash_contract_call ` function only returns ContractCall events and my reasoning was that fetching the ContractCall events in a tx_hash, that's pointed in the PollStarted event would be exactly what is needed.
What other ContractCall events that the code doesn't care about can originate in a single transaction 🤔